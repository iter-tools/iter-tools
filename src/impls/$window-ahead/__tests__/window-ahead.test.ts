/**
 * @generated-from ./$window-ahead.test.ts
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

import { windowAhead } from 'iter-tools-es';
import { wrap, unwrapDeep } from '../../../test/helpers.js';

describe('windowAhead', () => {
  describe('when source is empty', () => {
    it('yields no windows', () => {
      expect(unwrapDeep(windowAhead({ filler: 0 }, 3, null))).toEqual([]);
      expect(unwrapDeep(windowAhead({ filler: 0 }, 3, undefined))).toEqual([]);
      expect(unwrapDeep(windowAhead({ filler: 0 }, 3, wrap([])))).toEqual([]);
    });
  });

  describe('when size(source) < size', () => {
    it('yields only partial windows', () => {
      expect(unwrapDeep(windowAhead({ filler: 0 }, 3, wrap([1, 2])))).toEqual([
        [1, 2, 0],
        [2, 0, 0],
      ]);
    });
  });

  describe('when size(source) === size', () => {
    it('yields one full window, then partial windows', () => {
      expect(unwrapDeep(windowAhead({ filler: 0 }, 3, wrap([1, 2, 3])))).toEqual([
        [1, 2, 3],
        [2, 3, 0],
        [3, 0, 0],
      ]);
    });
  });

  describe('when size(source) > size', () => {
    it('yields size(source)-size full windows, then partial windows', () => {
      expect(unwrapDeep(windowAhead({ filler: 0 }, 2, wrap([1, 2, 3])))).toEqual([
        [1, 2],
        [2, 3],
        [3, 0],
      ]);
    });
  });

  describe('when useFiller is false', () => {
    it('yields shorter windows instead of windows with filler', () => {
      expect(unwrapDeep(windowAhead({ useFiller: false }, 3, wrap([1, 2, 3])))).toEqual([
        [1, 2, 3],
        [2, 3],
        [3],
      ]);
    });
  });

  it('has a default filler of undefined', () => {
    expect(unwrapDeep(windowAhead(2, wrap([1])))).toEqual([[1, undefined]]);
  });

  describe('when arguments are invalid', () => {
    it('throws', () => {
      expect(() => windowAhead('foo' as any, wrap([1, 2, 3]))).toThrowErrorMatchingSnapshot();
    });
  });
});
