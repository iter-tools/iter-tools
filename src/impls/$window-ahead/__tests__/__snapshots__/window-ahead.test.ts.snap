// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`windowAhead when arguments are invalid throws 1`] = `"windowAhead must be passed a numeric size"`;
